File: src/main/java/demo/kotlin/web/ApiRoutes.java
package demo.kotlin.web;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.reactive.function.server.RouterFunction;
import org.springframework.web.reactive.function.server.ServerResponse;

import static org.springframework.http.MediaType.APPLICATION_JSON;
import static org.springframework.web.reactive.function.server.RequestPredicates.GET;
import static org.springframework.web.reactive.function.server.RequestPredicates.accept;
import static org.springframework.web.reactive.function.server.RequestPredicates.path;
import static org.springframework.web.reactive.function.server.RouterFunctions.nest;
import static org.springframework.web.reactive.function.server.RouterFunctions.route;

@Configuration
public class ApiRoutes {
    private CowHandler cowHandler;

    public ApiRoutes(CowHandler cowHandler) {
        this.cowHandler = cowHandler;
    }

    @Bean
    public RouterFunction<ServerResponse> appRouter() {
        return nest(accept(APPLICATION_JSON),
                nest(path("/api/cows"),
                        route(GET("/{name}"), cowHandler::findByName)
                        .andRoute(GET("/"), cowHandler::findAll)));
    }
}

File: src/main/kotlin/demo/kotlin/web/ApiRoutes.kt
package demo.kotlin.web

import demo.kotlin.web.handlers.AuthenticationHandler
import demo.kotlin.web.handlers.CowHandler
import demo.kotlin.web.handlers.UserHandler
import demo.kotlin.web.handlers.save
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.http.MediaType.APPLICATION_JSON
import org.springframework.web.reactive.function.server.router

@Configuration
class ApiRoutes(
        private val cowHandler: CowHandler,
        private val authenticationHandler: AuthenticationHandler,
        private val userHandler: UserHandler
) {

    @Bean
    fun appRouter() = router {
        accept(APPLICATION_JSON).nest {
            "/api".nest {
                "/cows".nest {
                    GET("/{id}", cowHandler::findById)
                    GET("/name/{name}", cowHandler::findByName)
                    GET("/", cowHandler::findAll)
                    POST("/") { cowHandler.save(it) }
                }
                "/users".nest {
                    GET("/{id}", userHandler::findById)
                    DELETE("/{id}", userHandler::deleteById)
                    POST("/") { userHandler.save(it) }
                }
            }
            "/auth".nest {
                POST("/", authenticationHandler::auth)
            }
        }
    }
}
